### PR TITLE
api inference mini fork

### PR DESCRIPTION
**Pull request by oOraph**
Fork of the toolkit carrying the bits HF Inference needs: widget/api-inference response compatibility, multi-task deployments, and a much smaller idle memory footprint so we can overcommit nodes.

Everything new is opt-in through env vars, so the default behaviour of the image is unchanged unless stated otherwise.

## Inference / API behaviour

- **Multi task support with one deployment** (example: `sentence-similarity` + `sentence-embeddings`): pipelines are loaded lazily per task and cached, plus a `/pipeline/<task>` route for api-inference backward compat (`API_INFERENCE_COMPAT`).
- **api-inference response compat** (`API_INFERENCE_COMPAT=true`): response remapping for `text-classification`, `token-classification`, `zero-shot-classification`, `image-segmentation` (default score), `feature-extraction` (drops the batch-size dim), and `/pipeline/sentence-embeddings` substituted for `/pipeline/feature-extraction` on sentence-transformers models. Defaults for `DEFAULT_TOP_K` and `DEFAULT_AGGREGATION_STRATEGY`.
- **Overridable default inference params** so the default inference is sane when the user does not specify any: `DEFAULT_NUM_INFERENCE_STEPS` and `DEFAULT_GUIDANCE_SCALE` (diffusion), `DEFAULT_CONTENT_TYPE`, `DEFAULT_ACCEPT`.
- **Content-type / accept handling**: case-insensitive, tolerates extra params (`; charset=...`) instead of requiring an exact match, accepts `application/octet-stream`, and raises an explicit error instead of silently returning no (de)serializer or no content-type.
- **`IGNORE_CUSTOM_HANDLER`**: disable model-provided custom handlers (used on HF Inference).

## Memory footprint / overcommit

- **Idle unload** (`UNLOAD_IDLE`, `IDLE_TIMEOUT`): uvicorn replaced by gunicorn so an idle asgi worker can suppress itself and be respawned by the arbiter — the most immediate way we found to release memory without digging into the various lib / allocator mechanics. Model download and pipeline load become lazy (first request pays a slightly slower cold start).
- **Lazy imports of memory greedy libs** (`transformers`, `diffusers`, `sentence_transformers`) via the new `heavy_utils` module, so an idle worker keeps the minimum resident.
- **`DISCARD_LEFT`**: drop a request if the caller is no longer waiting for the answer, to relieve pressure during bursts. Requires the proxy to close the connection to be effective.
- **Adaptive latency guard** (`LATENCY_GUARD_ENABLED`, opt-in): dual EMA on pure inference latency (fast ~3-5 req vs slow ~20 req baseline); rejects with 503 while recent latency exceeds `LATENCY_OVERLOAD_FACTOR` × baseline, recovers below `LATENCY_RECOVERY_FACTOR` × baseline. Freezing is time bounded by `LATENCY_FREEZE_SECONDS` so the guard re-probes instead of staying frozen forever. Exposed as `inf_accepting` / `inf_auto_frozen` on `/metrics`.
- **Graceful shutdown**: on SIGTERM/SIGINT the worker drains in-flight and queued requests before leaving (`GRACEFUL_TIMEOUT`, `WORKERS`).

## Dependency upgrade (transformers 5.x / torch 2.6)

`nvidia/Nemotron-3-Embed-1B-BF16` failed on HF Inference because its architecture (`model_type` `ministral3`, added in transformers 5.0.0) is not recognized by transformers 4.51.3. The whole ML stack moves to the latest mutually-compatible set:

| | from | to |
|---|---|---|
| transformers | 4.51.3 | 5.14.1 |
| huggingface_hub | 0.30.2 | 1.24.0 |
| sentence_transformers | 4.0.2 | 5.6.0 |
| diffusers | 0.33.1 | 0.39.0 |
| accelerate | 1.6.0 | 1.14.0 |
| peft | 0.15.1 | 0.19.1 |
| torch | 2.5.1 | 2.6.0 |

- `huggingface_hub` >= 1.5.0 is required by transformers 5.x; the removed `hf_transfer` extra is dropped, `hf_xet` kept. `huggingface-cli` became a stub that exits 1 in hub 1.x, so the entrypoint uses `hf` instead.
- `torchvision` / `torchaudio` are pinned to the releases matching torch 2.6 (0.21.0 / 2.6.0). Left unpinned, pip resolves a CUDA 13 `torchaudio` while torch is a CUDA 12 build, and since transformers 5.x imports `torchaudio` unconditionally, every model load dies on `libcudart.so.13`.
- torch 2.6 is required because transformers 5.x refuses to `torch.load` `.bin` checkpoints with torch < 2.6 (CVE-2025-32434).
- diffusers 0.39 because 0.36 still imports the removed `MT5Tokenizer`.
- **TensorFlow is gone from transformers v5**, so TF support, the TF code paths and the TF tests are removed (the image has always been torch-only in practice). `require_tf` no longer exists in `transformers.testing_utils`, so the TF tests could not even be imported.
- transformers v5 dropped the `framework` pipeline arg; it is no longer forwarded.
- Starlette 1.0 removed `on_startup`/`on_shutdown`, so startup/shutdown moved to the ASGI lifespan protocol. Pinned `starlette>=1.0.1` for GHSA-86qp-5c8j-p5mr (Host header poisons `request.url.path`).
- Added `timm` (object detection) and `hf_xet` (xethub support).

### Four pipeline tasks had to be brought back

transformers 5 **deleted** the `question-answering`, `summarization`, `translation` and `text2text-generation` pipelines (`MIGRATION_GUIDE_V5.md`) — the classes and their modules are gone from the package and from `main`, so there is nothing left to re-register. Upstream's suggestion is to use a chat model with `TextGenerationPipeline`, which is advice for *choosing* a model and not something a serving layer can follow: we have to run the seq2seq and QA models that already exist on the Hub.

Routing them to `text-generation` is not a workaround either — encoder-decoder models are rejected outright (T5, Marian: `AutoModelForCausalLM` only), and BART is *silently* loaded as decoder-only with the whole encoder discarded, i.e. fluent nonsense instead of an error.

Everything those pipelines rely on still ships in v5 (`AutoModelForSeq2SeqLM`, `AutoModelForQuestionAnswering`, `SquadExample`, `squad_convert_examples_to_features`, `ChunkPipeline`, …), so the two upstream modules are vendored from the v4.51.3 tag under `vendor/` and dispatched from `legacy_transformers_utils`, exactly the way the sentence-transformers and diffusers tasks already are. Vendoring rather than reimplementing keeps upstream's edge-case handling: QA span extraction with offset mapping and `doc_stride` windowing, min/max length validation, translation language-pair parsing. Deviations from upstream are minimal and listed in each module header.

Two integration fixtures also had to be replaced because transformers 5 cannot load them at all: the image-segmentation model has `"vocab_size": {}` in its config (rejected by the stricter config validation), and the table-question-answering one needs the removed `TapexTokenizer`.

## Build

- Dockerfile split into a `builder` stage (compilers, kenlm, deps into a venv) and a lean `runtime` stage that only takes the venv and the sources — no compilers or dev headers in the shipped image.
- Requirements split from the source code layer to optimize build time and layer reuse, with a `REQUIREMENTS` build arg; `requirements.txt.cpu` builds a light CPU image (`make inference-pytorch-cpu`) off the CPU wheel index.

## Misc

- `LOG_LEVEL` env var.
- Fixed several async bugs: long blocking sync calls (model download) wrapped in a thread, coroutine/threading mixing causing blocking, and the idle check now covers the whole predict call so a worker cannot be kicked mid-request.

### CODE REVIEW

File: src/huggingface_inference_toolkit/diffusers_utils.py [hunk 1]
```diff
@@ -63,6 +64,16 @@ def __call__(
             kwargs.pop("num_images_per_prompt")
             logger.warning("Sending num_images_per_prompt > 1 to pipeline is not supported. Using default value 1.")
 
+        if "num_inference_steps" not in kwargs:
+            default_num_steps = os.environ.get("DEFAULT_NUM_INFERENCE_STEPS")
+            if default_num_steps:
+                kwargs["num_inference_steps"] = int(default_num_steps)
+
```
Comment (oOraph) [inline src/huggingface_inference_toolkit/diffusers_utils.py:71]: useful for sd 3.5 turbo for example -> we don't wan't a guidance scale >=0 if num inference steps is too low

File: src/huggingface_inference_toolkit/env_utils.py [hunk 2]
```diff
@@ -20,3 +23,7 @@ def strtobool(val: str) -> bool:
     raise ValueError(
         f"Invalid truth value, it should be a string but {val} was provided instead."
     )
+
+
+def api_inference_compat():
```
Comment (oOraph) [inline src/huggingface_inference_toolkit/env_utils.py:28]: with this env var we intend to handle the small response differences between the api inference widgets on the hub and on endpoints ui. TODO: we should probably unify both widgets instead

File: src/huggingface_inference_toolkit/webservice_starlette.py [hunk 3]
```diff
@@ -148,14 +166,19 @@ async def predict(request):
...
     )
 else:
+    routes = [
+        Route("/", health, methods=["GET"]),
+        Route("/health", health, methods=["GET"]),
+        Route("/", predict, methods=["POST"]),
+        Route("/predict", predict, methods=["POST"]),
+        Route("/metrics", metrics, methods=["GET"]),
+    ]
+    if api_inference_compat():
```
Comment (oOraph) [inline src/huggingface_inference_toolkit/webservice_starlette.py:176]: I only activated multi task for api inference but we may want to remove this condition and just always support it

Comment (oOraph) [reply src/huggingface_inference_toolkit/webservice_starlette.py:176]: actually thinking about it -> we may want a separate env var (and keep deactivated by default for regular users, and provide an option for it in endpoints instead) because the pod may consume more ram than expected (due to the pipeline duplications) with this route
